torchdynamo/spec.py

```diff
@@ -0,0 +1,174 @@
+                spec.add_element(_type_to_element_close(t))
+            if isinstance(t, dict):
+                spec.add_element(_type_to_element_open(t))
+                for k, v in t:
+                    # TODO(voz): Handle keys properly, this has a chance of causing a false-negative on export match.
+                    _describe_spec(v)
+                spec.add_element(_type_to_element_close(t))
+
+        _describe_spec(t)
+        return spec
+
+    def __init__(self):
+        self.elements = []
+        self.tensors = 0
+
+    @unique
+    class Element(str, Enum):
+        TENSOR = 0
+        OPEN_LIST = 1
+        CLOSE_LIST = 2
+        OPEN_TUPLE = 3
+        CLOSE_TUPLE = 4
+        OPEN_DICT = 5
+        CLOSE_DICT = 6
+        # TODO(voz): Exhaustively support all spec elements
+
+        @classmethod
+        def _repr_dict(self):
+            return {
+                Spec.Element.TENSOR: "T",
+                Spec.Element.OPEN_LIST: "List[",
+                Spec.Element.CLOSE_LIST: "]",
+                Spec.Element.OPEN_TUPLE: "Tuple(",
+                Spec.Element.CLOSE_TUPLE: ")",
+                Spec.Element.OPEN_DICT: "Dict{",
+                Spec.Element.CLOSE_DICT: "}",
+            }
+
+        def __repr__(self):
+            return self._repr_dict()[self]
+
+    def add_element(self, element: Element):
+        # TODO(voz): Add validation logic, (ex: no list close if no list opens on the stack).
+        # It currently exists in apply, but not here.
+        self.elements.append(element)
+        if element == Spec.Element.TENSOR:
+            self.tensors += 1
+
+    def __repr__(self):
+        rep = ""
+        for x in self.elements:
+            rep += f"{x!r}"
+        return rep
+
+    @staticmethod
+    def apply_spec(spec, tensors: List[torch.Tensor]):
+        r"""
+        Given a spec and a list of tensors, return a fully materialized data structure containing
+        all the Tensors.
+
+        Note: If the spec starts with a collection, the result will mirror it. Specs of only tensors, or starting
+        With tensors will be added to a Tuple
+        """
+        if len(tensors) == 0:
+            return []
+
+        assert spec.tensors == len(
+            tensors
+        ), f"Spec length {spec.elements} does not match tensor list length {len(tensors)}"
+
+        reversed_elements = spec.elements[::-1]
+        collection_stack = []
+        current_collection = None
+        idx = 0
+
+        def _append(e, collection):
+            if isinstance(collection, list):
+                collection.append(e)
+            elif isinstance(collection, tuple):
+                # TODO(voz): Optimize with forward scanning, this isnt that cheap
+                collection += (e,)
+            else:
+                assert False, f"Unsupported collection type {collection.__class__}"
+
+        result = None
+        current_element = reversed_elements.pop()
+        while current_element is not None:
```

Comment:
At least for what we have so far, this could totally just be a `pytree.TreeSpec()`.  The features supported by a pytees are a superset.

Reply:
Does pytree offer the ability to build it up iteratively, as we do with Spec? That's the biggest gap here, so far.